metadata-ingestion/src/datahub/ingestion/source/aerospike.py

```diff
@@ -0,0 +1,591 @@
+                yield from [
+                    mcp.as_workunit()
+                    for mcp in MetadataChangeProposalWrapper.construct_many(
+                        entityUrn=dataset_urn.urn(),
+                        aspects=[
+                            schema_metadata,
+                            dataset_properties,
+                            data_platform_instance,
+                        ],
+                    )
+                ]
+
+    def _infer_schema_metadata(
+        self,
+        dataset_urn: DatasetUrn,
+        as_set: AerospikeSet,
+        dataset_properties: DatasetPropertiesClass,
+    ) -> SchemaMetadata:
+        set_schema: Dict[
+            Tuple[str, ...], SchemaDescription
+        ] = construct_schema_aerospike(
+            client=self.aerospike_client,
+            as_set=as_set,
+            delimiter=".",
+            records_per_second=self.config.records_per_second,
+            sample_size=self.config.schemaSamplingSize,
+        )
+
+        if self.config.inferSchemaDepth != -1:
+            # Infer schema only at the bins level
+            set_schema = {
+                k: v
+                for k, v in set_schema.items()
+                if len(k) <= self.config.inferSchemaDepth
+            }
+
+        # initialize the schema for the set
+        canonical_schema: List[SchemaField] = []
+        max_schema_size = self.config.maxSchemaSize
+        set_schema_size = len(set_schema.values())
+        set_fields: Union[
+            List[SchemaDescription], ValuesView[SchemaDescription]
+        ] = set_schema.values()
+        assert max_schema_size is not None
+        if set_schema_size > max_schema_size:
+            # downsample the schema, using frequency as the sort key
+            self.report.report_warning(
+                title="Too many schema fields",
+                message=f"Downsampling the collection schema because it has too many schema fields. Configured threshold is {max_schema_size}",
+                context=f"Schema Size: {set_schema_size}, Collection: {dataset_urn}",
+            )
+            # Add this information to the custom properties so user can know they are looking at downsampled schema
+            dataset_properties.customProperties["schema.downsampled"] = "True"
+            dataset_properties.customProperties[
+                "schema.totalFields"
+            ] = f"{set_schema_size}"
+
+        logger.debug(f"Size of set fields = {len(set_fields)}")
+        # append each schema field (sort so output is consistent)
+        for schema_field in sorted(
+            set_fields,
+            key=lambda x: (
+                -x["count"],
+                x["delimited_name"],
+            ),  # Negate `count` for descending order, `delimited_name` stays the same for ascending
+        )[0:max_schema_size]:
+            field = SchemaField(
+                fieldPath=schema_field["delimited_name"],
```

Comment:
are these fields guaranteed to be present? do we need more defensive coding here?